misc/python/materialize/mzbuild.py

```diff
@@ -514,24 +505,33 @@ def build(self) -> None:
         ]
         spawn.runv(cmd, stdin=f, stdout=sys.stderr.buffer)
 
-    def acquire(self) -> AcquiredFrom:
-        """Download or build the image.
-
-        Returns:
-            acquired_from: How the image was acquired.
-        """
+    def acquire(self) -> None:
+        """Download or build the image if it does not exist locally."""
+        if self.acquired:
+            return
 
+        ui.header(f"Acquiring {self.spec()}")
         try:
             spawn.runv(
                 ["docker", "pull", self.spec()],
                 stdout=sys.stderr.buffer,
             )
-            return AcquiredFrom.REGISTRY
         except subprocess.CalledProcessError:
-            pass
+            for dep in self.dependencies.values():
+                dep.acquire()
+            self.build()
+        self.acquired = True
 
-        self.build()
-        return AcquiredFrom.LOCAL_BUILD
+    def ensure(self) -> None:
+        """Ensure the image exists on Docker Hub if it is publishable.
+
+        The image is pushed using its spec as its tag.
+        """
+        if self.publish and is_docker_image_pushed(self.spec()):
+            ui.say(f"{self.spec()} already exists")
+            return
+        self.acquire()
```

Comment:
This has the annoying property that `def aquire()` will attempt to pull an image we just checked it doesn't exist. I think this call should be `self.build()` instead, and in there we need to do the loop that is currently in acquire:

```python
            for dep in self.dependencies.values():
                dep.acquire()
```

Reply:
I need to get more sleep. That's another great suggestion.